File: server/models/automation.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, Text
from server.core.database import Base
from datetime import datetime

class AutomationTask(Base):
    __tablename__ = "automation_tasks"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(String, index=True, nullable=True)
    name = Column(String, nullable=True)
    agent_id = Column(String, nullable=False)
    input_prompt = Column(Text, nullable=True)
    
    # Timing configuration
    # If cron_expression is present, it takes precedence over interval_seconds
    cron_expression = Column(String, nullable=True)  # e.g. "0 12 * * *"
    interval_seconds = Column(Integer, nullable=True)
    
    is_one_time = Column(Boolean, default=False)
    is_active = Column(Boolean, default=True)
    
    next_run_time = Column(DateTime, nullable=True)
    last_run_time = Column(DateTime, nullable=True)
    
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class AutomationLog(Base):
    __tablename__ = "automation_logs"

    id = Column(Integer, primary_key=True, index=True)
    task_id = Column(Integer, index=True)
    status = Column(String)  # RUNNING, SUCCESS, FAILURE
    output = Column(Text, nullable=True)
    error_message = Column(Text, nullable=True)
    start_time = Column(DateTime, default=datetime.utcnow)
    end_time = Column(DateTime, nullable=True)

File: admin/src/api/automation.ts
import request from './axios'

export interface AutomationTask {
  id: number
  name: string
  agent_id: string
  input_prompt: string
  cron_expression?: string
  interval_seconds?: number
  is_one_time: boolean
  is_active: boolean
  last_run_time?: string
  created_at: string
}

export interface AutomationLog {
  id: number
  task_id: number
  status: string
  output?: string
  error_message?: string
  start_time: string
  end_time?: string
}

export const getTasks = () => {
  return request.get<AutomationTask[]>('/api/v1/automation/tasks')
}

export const createTask = (data: any) => {
  return request.post<AutomationTask>('/api/v1/automation/tasks', data)
}

export const updateTask = (id: number, data: any) => {
  return request.put<AutomationTask>(`/api/v1/automation/tasks/${id}`, data)
}

export const deleteTask = (id: number) => {
  return request.delete(`/api/v1/automation/tasks/${id}`)
}

export const runTask = (id: number) => {
  return request.post(`/api/v1/automation/tasks/${id}/run`)
}

export const getTaskLogs = (id: number) => {
  return request.get<AutomationLog[]>(`/api/v1/automation/tasks/${id}/logs`)
}

export const generateCronExpression = (description: string) => {
  return request.post<{ cron_expression: string }>('/api/v1/automation/cron/generate', { description })
}
